File: src/routes/customer-image.routes.ts
import { Hono } from "hono";
import type { Env } from "@/types/hono";
import { CustomerImageService, type UploadImageInput } from "@/services/customer-image.service";
import { successResponse } from "@/utils/response";
import { authMiddleware } from "@/middleware/auth";
import { toResponse, toResponseArray } from "@/utils/response-transformer";
import { z } from "zod";
import type { CustomerImageMetadata } from "@/types/customer/image/customer_image";

export const customerImageRoutes = new Hono<Env>();

let service: CustomerImageService | null = null;

function getService(): CustomerImageService {
  if (!service) {
    service = new CustomerImageService();
  }
  return service;
}

// Apply auth middleware
customerImageRoutes.use("*", authMiddleware);

// GET /api/customers/images - List all images (no customerId filter)
customerImageRoutes.get("/images", async (c) => {
  const user = c.get("user");

  const images = await getService().listAllImages(user.companyId!);
  return c.json(successResponse(toResponseArray(images)));
});

// GET /api/customers/images/:imageId/download - Download image binary by imageId only
customerImageRoutes.get("/images/:imageId/download", async (c) => {
  const user = c.get("user");
  const imageId = c.req.param("imageId");

  const { buffer, metadata } = await getService().getImageData(user.companyId!, imageId);

  return new Response(buffer, {
    headers: {
      "Content-Type": metadata.mimeType,
      "Content-Length": metadata.size.toString(),
      "Content-Disposition": `attachment; filename*=UTF-8''${encodeURIComponent(metadata.filename)}`,
      "Access-Control-Allow-Origin": "*",
      "Access-Control-Allow-Methods": "GET, OPTIONS",
      "Access-Control-Allow-Headers": "Content-Type, Authorization",
    },
  });
});

// GET /api/customers/:customerId/images - List images for customer (metadata only)
customerImageRoutes.get("/:customerId/images", async (c) => {
  const user = c.get("user");
  const customerId = c.req.param("customerId");

  const images = await getService().listImagesByCustomer(user.companyId!, customerId);
  return c.json(successResponse(toResponseArray(images)));
});

// POST /api/customers/images/by-labels - Get images by label IDs
const byLabelsSchema = z.object({
  labelIds: z.array(z.string()).default([]),
});

customerImageRoutes.post("/images/by-labels", async (c) => {
  const user = c.get("user");
  const body = await c.req.json();
  const { labelIds } = byLabelsSchema.parse(body);

  const images = labelIds.length === 0
    ? await getService().listAllImages(user.companyId!)
    : await getService().getImagesByLabels(user.companyId!, labelIds);
  return c.json(successResponse(toResponseArray(images)));
});

// GET /api/customers/images/:imageId - Get image metadata
customerImageRoutes.get("/images/:imageId", async (c) => {
  const user = c.get("user");
  const imageId = c.req.param("imageId");

  const image = await getService().getImageMetadata(user.companyId!, imageId);
  return c.json(successResponse(toResponse(image)));
});

// Shared upload handler
async function handleImageUpload(c: any, customerId?: string) {
  const user = c.get("user");

  const formData = await c.req.formData();
  const files = formData.getAll("files") as File[];
  const description = formData.get("description") as string || "";
  const labelsStr = formData.get("labels") as string || "[]";

  let labels: string[] = [];
  try {
    labels = JSON.parse(labelsStr);
  } catch {
    labels = [];
  }

  if (files.length === 0) {
    return c.json({ success: false, error: { message: "No files provided" } }, 400);
  }

  const uploads: UploadImageInput[] = await Promise.all(
    files.map(async (file) => ({
      title: file.name, // Always use filename as title
      description,
      filename: file.name,
      mimeType: file.type,
      data: Buffer.from(await file.arrayBuffer()),
      labels,
    }))
  );

  if (uploads.length === 1) {
    const image = await getService().uploadImage(
      user.companyId!,
      customerId,
      user._id?.toString()!,
      uploads[0]!
    );
    return c.json(successResponse(toResponse(image)), 201);
  } else {
    const images = await getService().uploadMultipleImages(
      user.companyId!,
      customerId,
      user._id?.toString()!,
      uploads
    );
    return c.json(successResponse(toResponseArray(images)), 201);
  }
}

// POST /api/customers/images - Upload image(s) without customerId
customerImageRoutes.post("/images", async (c) => {
  return handleImageUpload(c);
});

// POST /api/customers/:customerId/images - Upload image(s) with customerId
customerImageRoutes.post("/:customerId/images", async (c) => {
  const customerId = c.req.param("customerId");
  return handleImageUpload(c, customerId);
});

// PUT /api/customers/images/:imageId - Update image metadata
customerImageRoutes.put("/images/:imageId", async (c) => {
  const user = c.get("user");
  const imageId = c.req.param("imageId");
  const body = await c.req.json();

  const updateSchema = z.object({
    title: z.string().optional(),
    description: z.string().optional(),
    labels: z.array(z.string()).optional(),
  });

  const input = updateSchema.parse(body);
  const image: CustomerImageMetadata = await getService().updateImage(user.companyId!, imageId, input);

  return c.json(successResponse(toResponse(image)));
});

// DELETE /api/customers/images/:imageId - Delete image by imageId only
customerImageRoutes.delete("/images/:imageId", async (c) => {
  const user = c.get("user");
  const imageId = c.req.param("imageId");

  await getService().deleteImage(user.companyId!, imageId);

  return c.json(successResponse({ message: "Image deleted successfully" }));
});

// Resumable Upload Endpoints

// POST /api/customers/:customerId/upload/init - Initialize resumable upload
customerImageRoutes.post("/:customerId/upload/init", async (c) => {
  const user = c.get("user");
  const customerId = c.req.param("customerId");
  const body = await c.req.json();

  const initSchema = z.object({
    filename: z.string(),
    mimeType: z.string(),
    totalSize: z.number().int().positive(),
  });

  const input = initSchema.parse(body);
  const result = await getService().initUpload(user.companyId!, customerId, user._id?.toString()!, input);

  return c.json(successResponse(result));
});

// POST /api/customers/:customerId/upload/chunk - Upload a chunk
customerImageRoutes.post("/:customerId/upload/chunk", async (c) => {
  const user = c.get("user");
  const customerId = c.req.param("customerId");
  const uploadId = c.req.query("uploadId");
  const chunkIndex = c.req.query("index");

  if (!uploadId || !chunkIndex) {
    return c.json({ success: false, error: { message: "Missing uploadId or index" } }, 400);
  }

  // Get raw binary body
  const buffer = await c.req.arrayBuffer();
  if (!buffer || buffer.byteLength === 0) {
    return c.json({ success: false, error: { message: "Empty chunk data" } }, 400);
  }

  const result = await getService().uploadChunk(
    user.companyId!,
    uploadId,
    parseInt(chunkIndex),
    Buffer.from(buffer)
  );

  return c.json(successResponse(result));
});

// POST /api/customers/:customerId/upload/finalize - Finalize upload
customerImageRoutes.post("/:customerId/upload/finalize", async (c) => {
  const user = c.get("user");
  const customerId = c.req.param("customerId");
  const body = await c.req.json();

  const finalizeSchema = z.object({
    uploadId: z.string(),
    title: z.string(),
    description: z.string().optional(),
    labels: z.array(z.string()).optional(),
  });

  const input = finalizeSchema.parse(body);
  const image = await getService().finalizeUpload(user.companyId!, input.uploadId, {
    title: input.title,
    description: input.description,
    labels: input.labels,
  });

  return c.json(successResponse(toResponse(image)), 201);
});

File: fisoft/src/main/kotlin/routes/CustomerImageRoutes.kt
package com.ilyasdemirkiran.routes

import com.ilyasdemirkiran.middleware.authenticateUser
import com.ilyasdemirkiran.middleware.currentCompanyId
import com.ilyasdemirkiran.services.CustomerImageService
import com.ilyasdemirkiran.utils.successResponse
import io.ktor.http.*
import io.ktor.http.content.*
import io.ktor.server.application.*
import io.ktor.server.request.*
import io.ktor.server.response.*
import io.ktor.server.routing.*
import kotlinx.serialization.Serializable

@Serializable
data class ByLabelsReq(val labelIds: List<String> = emptyList())

@Serializable
data class UpdateImageReq(val labels: List<String> = emptyList())

fun Route.customerImageRoutes() {
    val imageService = CustomerImageService()

    route("/api/customers/images") {
        // GET /api/customers/images - List all images across company
        get {
            call.authenticateUser()
            val companyId = call.currentCompanyId
            val images = imageService.listAllImages(companyId)
            call.respond(HttpStatusCode.OK, successResponse(images))
        }

        // POST /api/customers/images/by-labels - Filter by labels
        post("/by-labels") {
            call.authenticateUser()
            val companyId = call.currentCompanyId
            val req = call.receive<ByLabelsReq>()
            val images = imageService.getImagesByLabels(companyId, req.labelIds)
            call.respond(HttpStatusCode.OK, successResponse(images))
        }

        // GET /api/customers/images/{imageId} - Get image metadata
        get("/{imageId}") {
            call.authenticateUser()
            val companyId = call.currentCompanyId
            val imageId = call.parameters["imageId"] ?: ""
            val image = imageService.getImageMetadata(companyId, imageId)
            call.respond(HttpStatusCode.OK, successResponse(image))
        }

        // GET /api/customers/images/{imageId}/download - Raw image download
        get("/{imageId}/download") {
            call.authenticateUser()
            val companyId = call.currentCompanyId
            val imageId = call.parameters["imageId"] ?: ""
            val (bytes, mimeType) = imageService.downloadImage(companyId, imageId)
            call.response.header(
                HttpHeaders.ContentDisposition,
                ContentDisposition.Attachment.withParameter(ContentDisposition.Parameters.FileName, "image.jpg").toString()
            )
            call.respondBytes(bytes, ContentType.parse(mimeType))
        }

        // POST /api/customers/images - Upload image without customerId
        post {
            call.authenticateUser()
            val companyId = call.currentCompanyId
            val multipart = call.receiveMultipart()
            var filename = "image.jpg"
            var mimeType = "image/jpeg"
            var bytes: ByteArray? = null
            var labels: List<String> = emptyList()

            multipart.forEachPart { part ->
                when (part) {
                    is PartData.FileItem -> {
                        filename = part.originalFileName ?: "image.jpg"
                        mimeType = part.contentType?.toString() ?: "image/jpeg"
                        bytes = part.streamProvider().readBytes()
                    }
                    is PartData.FormItem -> {
                        if (part.name == "labels") {
                            labels = part.value.split(",").map { it.trim() }.filter { it.isNotEmpty() }
                        }
                    }
                    else -> {}
                }
                part.dispose()
            }

            if (bytes == null) {
                call.respond(HttpStatusCode.BadRequest, mapOf("success" to false, "error" to "No file provided"))
                return@post
            }

            val image = imageService.uploadImage(companyId, null, filename, bytes!!, mimeType, labels)
            call.respond(HttpStatusCode.Created, successResponse(image))
        }

        // PUT /api/customers/images/{imageId} - Update labels metadata
        put("/{imageId}") {
            call.authenticateUser()
            val companyId = call.currentCompanyId
            val imageId = call.parameters["imageId"] ?: ""
            val req = call.receive<UpdateImageReq>()
            val updated = imageService.updateImage(companyId, imageId, req.labels)
            call.respond(HttpStatusCode.OK, successResponse(updated))
        }

        // DELETE /api/customers/images/{imageId} - Delete image
        delete("/{imageId}") {
            call.authenticateUser()
            val companyId = call.currentCompanyId
            val imageId = call.parameters["imageId"] ?: ""
            imageService.deleteImage(companyId, imageId)
            call.respond(HttpStatusCode.OK, successResponse(mapOf("message" to "Image deleted successfully")))
        }
    }
}
